src/daemon/rpmostreed-os-experimental.cxx

```diff
@@ -205,20 +207,32 @@ osexperimental_handle_download_packages (RPMOSTreeOSExperimental *interface,
   if (!rpmostree_context_download_metadata (ctx, DNF_CONTEXT_SETUP_SACK_FLAG_SKIP_RPMDB,
                                             cancellable, &local_error))
     goto out;
-
   sack = dnf_context_get_sack (rpmostree_context_get_dnf (ctx));
-
+  RpmOstreeOverrideSourceKind source_kind;
+  source_name = rpmostree_parse_override_source (source, &source_kind, &local_error);
   for (const char* const* it = queries; it && *it; it++)
     {
-      auto query = static_cast<const char*> (*it);
-      g_autoptr(GPtrArray) pkglist = rpmostree_get_matching_packages (sack, query);
-
-      if (!pkglist || pkglist->len == 0)
-        {
-          glnx_throw (&local_error, "No matching packages for \"%s\"", query);
-          goto out;
-        }
-      g_ptr_array_add (dnf_pkgs, g_object_ref (pkglist->pdata[0]));
+       HyNevra nevra = NULL;
+       g_auto(HySubject) subject = hy_subject_create(static_cast<const char*> (*it));
+       hy_autoquery HyQuery query =
+         hy_subject_get_best_solution(subject, sack, NULL, &nevra, FALSE, TRUE, TRUE, TRUE, FALSE);
+       hy_query_filter (query, source_kind, HY_EQ, source_name);
+       g_autoptr(DnfPackageSet) pset = hy_query_run_set (query);
+
+       if (dnf_packageset_count (pset) == 0)
+       {
+         glnx_throw (&local_error, "No matches for '%s' in repo '%s'", (char*) it, source_name);
+         goto out;
+       }
+
+//      auto query = static_cast<const char*> (*it);
+//      g_autoptr(GPtrArray) pkglist = rpmostree_get_matching_packages (sack, query);
+//      if (!pkglist || pkglist->len == 0)
+//        {
+//          glnx_throw (&local_error, "No matching packages for \"%s\"", query);
+//          goto out;
+//        }
+      g_ptr_array_add (dnf_pkgs, g_object_ref (pset));
```

Comment:
I think this is the source of the `g_object_ref: assertion 'G_IS_OBJECT (object)' failed` assertion.

In the previous code, we were filtering down to exactly one package, so we were passing `g_object_ref (pkglist->pdata[0])`.  Here, are we trying to add in all of the packages?

OK, I think I see the problem (commented above).

Then to merge the arrays, you'll want to iterate over the child array and add each element to `dnf_pkgs`.